### ISSUE TITLE
grbl: serial port attach causes reset

### ISSUE THREAD

**Issue by JohnDMcMaster** [Labsmore/pyuscope#40]
TLDR: attaching a serial port in certain conditions causes GRBL to reset. I've worked around this in code for now by looking for a reset occurring during attach and printing a warning

Misc high level notes:
- Attaching a serial port after reset via cat (after stty) or pyserial reproduces
- Using screen once makes the issue go away on all future attaches
- Some stty options can make the board go beserk

Based on this I'm guessing its some sort of flow control option causing weird things to happen. But need more info

To reproduce :
- Unplug USB serial port
- Attach USB serial port
-stty -F /dev/ttyUSB0 115200 raw
- Run this several times
- cat /dev/ttyUSB0
- Each time you run it, despite expectation cat sends no data, will reset the controller ("Grbl 1.1f ['$' for help]")

Sample output after attaching serial port:

```
mcmaster@pop-os:~/doc/ext/pyuscope$ stty -F /dev/ttyUSB0 115200 raw
mcmaster@pop-os:~/doc/ext/pyuscope$ cat /dev/ttyUSB0

Grbl 1.1f ['$' for help]

Grbl 1.1f ['$' for help]
^C
mcmaster@pop-os:~/doc/ext/pyuscope$ cat /dev/ttyUSB0

Grbl 1.1f ['$' for help]
^C
mcmaster@pop-os:~/doc/ext/pyuscope$ cat /dev/ttyUSB0

Grbl 1.1f ['$' for help]
^C
```

Its believed the attach causes the issue as I also see the same issue from python and can't send commands at this time. 

To have grbl violently spew output:
- attach serial port
- stty -F /dev/ttyUSB0 115200
- cat /dev/ttyUSB0
- Run in terror

Sample output:

```
mcmaster@pop-os:~/doc/ext/pyuscope$ stty -F /dev/ttyUSB0 115200; cat /dev/ttyUSB0 |head -n 20


Grbl 1.1f ['$' for help]

ok

ok

ok

ok

error:2

ok

ok

ok
```

**Comment by JohnDMcMaster** [comment]
sample healthy open (ex: open with screen first):

```
mcmaster@pop-os:~/doc/ext/pyuscope$ ./test/grbl/probe.py --verbose
opening /dev/ttyUSB0
tx '?'
rx 60 bytes in 0.008 sec
00000000  3C 49 64 6C 65 7C 4D 50  6F 73 3A 30 2E 30 30 30  |<Idle|MPos:0.000|
00000010  2C 30 2E 30 30 30 2C 30  2E 30 30 30 7C 46 53 3A  |,0.000,0.000|FS:|
00000020  30 2C 30 7C 57 43 4F 3A  30 2E 30 30 30 2C 30 2E  |0,0|WCO:0.000,0.|
00000030  30 30 30 2C 30 2E 30 30  30 3E 0D 0A              |000,0.000>..    |
Open ok
```

sample workaround output:

```
mcmaster@pop-os:~/doc/ext/pyuscope$ ./test/grbl/probe.py --verbose
opening /dev/ttyUSB0
tx '?'
rx 0 bytes in 0.201 sec
grbl WARNING: failed to respond, attempting recovery
rx 0 bytes in 0.201 sec
rx 0 bytes in 0.201 sec
rx 0 bytes in 0.201 sec
rx 0 bytes in 0.201 sec
rx 0 bytes in 0.201 sec
rx 2 bytes in 0.165 sec
00000000  0D 0A                                             |..              |
rx 26 bytes in 0.000 sec
00000000  47 72 62 6C 20 31 2E 31  66 20 5B 27 24 27 20 66  |Grbl 1.1f ['$' f|
00000010  6F 72 20 68 65 6C 70 5D  0D 0A                    |or help]..      |
tx '?'
rx 60 bytes in 0.008 sec
00000000  3C 49 64 6C 65 7C 4D 50  6F 73 3A 30 2E 30 30 30  |<Idle|MPos:0.000|
00000010  2C 30 2E 30 30 30 2C 30  2E 30 30 30 7C 46 53 3A  |,0.000,0.000|FS:|
00000020  30 2C 30 7C 57 43 4F 3A  30 2E 30 30 30 2C 30 2E  |0,0|WCO:0.000,0.|
00000030  30 30 30 2C 30 2E 30 30  30 3E 0D 0A              |000,0.000>..    |
rx '<Idle|MPos:0.000,0.000,0.000|FS:0,0|WCO:0.000,0.000,0.000>'
grbl: recovered after 1.389 sec
Open ok
```

**Comment by JohnDMcMaster** [comment]
More info here: https://github.com/gnea/grbl/issues/160

**Comment by JohnDMcMaster** [comment]
https://github.com/pyserial/pyserial/issues/124

The termios workaround works for me. Integrating this fix into the code

**Comment by JohnDMcMaster** [comment]
Got a chance to test the termios workaround. A few notes:
- The first attach (this or even screen) brings grbl out of reset. So the recovery code is needed regardless
- I had my computer lock up once while testing this. Keep an eye on this

**Comment by JohnDMcMaster** [comment]
seems like this is as good as its going to be
